_posts/2024-10-08-Internal-functioning-of-sorting-in-hybrid-query.md

```diff
@@ -0,0 +1,122 @@
+
+The following example illustrates the sorting workflow in a hybrid query. In this case, a user triggers a hybrid query with two subqueries: `match` and `term`, and requests that the search results be ordered by stock price in descending order.
+
+First, the hybrid query executes the `match` and `term` subqueries separately, sorting each result set by the specified criterion—in this case, stock price. Unlike traditional sorting, which disregards relevance scores, hybrid queries include both the relevance score and the sorting fields in the results of query phase. These values are then used in the normalization process. During normalization, relevance scores are used to calculate normalization scores, which act as tie-breakers when subquery results are combined. Finally, the subquery results are combined at the coordinator level to form the final result, ensuring that the output is sorted by stock price in descending order, as requested.
+
+The following diagram illustrates the flow of the query phase and creating the final shard result based on sorting criteria.
+
+![hybrid-query-with-sorting](/assets/media/blog-images/2024-10-08-internal-functioning-of-sorting-in-hybrid-search/Hybrid-query-with-sorting.png){:class="img-centered"}
+
+
+## Detailed overview of sorting in a hybrid query 
+
+Building on the high-level logic discussed earlier, let's now dive into the detailed process of sorting in a hybrid query. This can be broken down into two key steps:
+
+1. Sorting individual subquery results at the shard level.
+2. Merging the multiple subquery results based on the sorting criteria at the coordinator node.
+
+Figure 1.2 illustrates these steps in detail. 
+
+![sorting-hld](/assets/media/blog-images/2024-10-08-internal-functioning-of-sorting-in-hybrid-search/Sorting-hld.png){:class="img-centered"}
+
+The coordinator node sends requests to data nodes to retrieve query results from the shards. Internally, during the query phase, the following process takes place:
+
+- A collector runs on each shard to gather results for each subquery. As results are collected, they are inserted into a priority queue that is ordered based on the sorting criteria.
+- At the end of the query phase, the subquery results are popped from the priority queue in sorted order.
+
+Once the coordinator node receives responses from all the data nodes, the normalization process begins. This process assigns a score to each subquery result and merges them into a final sorted list according to the specified sorting criteria. After normalization, the results move to the fetch phase, where the document IDs are used to retrieve the full document content from the shards. The final search response is then sent to the user.
+
```

Comment:
Above, last sentence: "the full document content" => either "the full document contents" or "the full documents"